README.md
[![UGOforNODE](https://raw.github.com/bluedge/gh-pages/master/images/UGOforNODE.png)](https://github.com/bluedge/UGOforNODE)
#UGOforNODE

Lightweight module for UNIX Symbolic notation "rwxrwxr--" to Octal notation "774" conversion.
[![Build Status](https://travis-ci.org/bluedge/UGOforNODE.png?branch=master)](https://travis-ci.org/bluedge)

## Install

```bash
npm install ugofornode
```

## Features List
	- toOctal from Symbolic notation to Octal notation
	- toSymbolic from Octal notation to Symbolic notation
	- getPermissionByClass return the Symbolic notation for a class (User, Group, Other)
	- isSymbolicValid check Symbolic notation validity
	- isOctalValid check the Octal notation validity
	
### Usage Example

```js	
	var UGOforNODE 			= require('ugofornode')
		, ugo 				= new UGOforNODE();
	
	// Get Octal notation from Array
	permission1 = ugo.toOctal( ["rwx", "rwx", "r--"]; // 774
	
	// Get Octal notation from String
	permission2 = ugo.toOctal( "rwxrwxr--" ); // 774
	
	// Get Symbolic notation (Array) from Octal
	permission3 = ugo.toSymbolic( 774 ); // ["rwx","rwx","r--"]
	
	// Get Symbolic notation (String) from Octal
	permission4 = ugo.toSymbolic( 774, "string" ); // "rwxrwxr--"
	
	// Get Symbolic notation of the specified class from Octal Notation 
	permission5 = ugo.getPermissionByClass( 774, 'OTHER' ); // "r--"
	
	// Get Symbolic notation of the specified class from Symbolic Notation (Array)
	permission6 = ugo.getPermissionByClass( ["rwx","rwx","r--"], 'OTHER' ); // "r--"
		
	// Validate Symbolic notation (Array)
	permission7 = ugo.isSymbolicValid( ["rwx","rwx","r--"] ); // true;
	
	// Validate Symbolic notation (String)
	permission8 = ugo.isSymbolicValid( "rwxrwxr--" ); // true;
	
	// Validate Octal notation
	permission9 = ugo.isOctalValid( 555 ); // true;
```

### Sponsors
UGOforNODE was created as part of a custom node development initiative sponsored by Bureaucloud (http://bureaucloud.com) & Bluedge(http://bluedge.co.uk) for a private application.
This simple and lean module quickly became so convenient for addressing permissions, we decided to share it with others.   

History.md
1.0.0 / 2013-01-22
==================
	* All Test passed on mocha
	* Ready for production version updated
	

0.3.0 / 2013-01-09
==================

	* Renamed methods
	* Removed factory pattern
	* isSymbolicValid() now accepts strings as a parameter
	

0.2.0 / 2012-12-12
==================

	* Add constant USER, GROUP, OTHER
	* Add @method getPermissionByClass
	* Add private method to convert UGO string notation to Array

  
0.1.0 / 2012-12-11
==================

	* Initial version  

test/index.js
var 	assert 			= require('assert')
	, 	UGOforNODE 		= require('../lib/main.js')
	, 	ugo				= new UGOforNODE();



describe( 'UGOforNODE', function() {

	// .toOctal
	describe( '.toOctal should convert Symbolic notation passed by String ["rwx", "rwx", "r--"] to Numeric notation', function() {
		it( "should return 774", function () {
			assert.strictEqual( 774, ugo.toOctal( ["rwx", "rwx", "r--"] ) );
		})
	});
	
	// .toOctal
	describe( '.toOctal should convert Symbolic notation passed by String "rwxrwxr--" to Numeric notation', function() {
		it( "should return 774", function () {
			assert.strictEqual( 774, ugo.toOctal( "rwxrwxr--" ) );
		})
	});
	
	// .toSymbolic
	describe( '.toSymbolic should convert Numeric notation 774 to Symbolic notation as array ["rwx", "rwx", "r--"]', function() {
		it( 'should return ["rwx","rwx","r--"]', function () {
			assert.deepEqual( ["rwx","rwx","r--"], ugo.toSymbolic( 774 ) );
		})
	});
	
	// .toSymbolic
	describe( '.toSymbolic should convert Numeric notation 774 to Symbolic notation as String "rwxrwxr--", option param "string"', function() {
		it( 'should return "rwxrwxr--"', function () {
			assert.strictEqual( "rwxrwxr--", ugo.toSymbolic( 774, "string" ) );
		})
	});
	
	// .getPermissionByClass
	describe( '.getPermissionByClass should return the Symbolic notation for the specified group, permission noted as Number', function() {
		it( 'should return "r--"', function () {
			assert.strictEqual( "r--", ugo.getPermissionByClass( 774, 'OTHER' ) );
		})
	});
	
	
	// .getPermissionByClass
	describe( '.getPermissionByClass should return the Symbolic notation for the specified group, permission noted as Array', function() {
		it( 'should return "r--"', function () {
			assert.strictEqual( "r--", ugo.getPermissionByClass( ["rwx","rwx","r--"], 'OTHER' ) );
		})
	});
	
	// .isSymbolicValid
	describe( '.isSymbolicValid validate Symbolic permission noted as Array', function() {
		it( 'should return true', function () {
			assert.strictEqual( true, ugo.isSymbolicValid( ["rwx","rwx","r--"] ) );
		})
	});
	
	// .isSymbolicValid
	describe( '.isSymbolicValid validate Symbolic permission noted as String', function() {
		it( 'should return true', function () {
			assert.strictEqual( true, ugo.isSymbolicValid( "rwxrwxr--" ) );
		})
	});
	
	// .isOctalValid
	describe( '.isOctalValid validate Octal', function() {
		it( 'should return true', function () {
			assert.strictEqual( true, ugo.isOctalValid( 555 ) );
		})
	});


});




//var ret = perm.toOctal( [ 'rwx', 'rwx', 'r--' ] );
//var ret = perm.toOctal( 'rwxrw-r--' );
//var ret2 = perm.toSymbolic( 764, 'string' );
//var ret3 = perm.getPermissionByClass( 777, 'OTHER' );
//console.log( ret );
//console.log( ret2 );
//console.log( 'ret3: ', ret3 );

//console.log( perm.isOctalValid( 777 ) );

lib/main.js
/**
 * UGOforNODE
 * Convert Symbolic notation of UNIX permission (rwx) to octal-like
 * notation (thru binary conversion) and vice-versa.
 * Note: UGO stands for to Unix denomination; User, Group, Other
 * Pattern: [Prototype]
 * Creation date: 2012/12/11
 * Async: False
 	* @class UGOforNODE
**/



// Constructor
var UGOforNODE = function () {};


// Export
module.exports = UGOforNODE;


/*** CONSTANTS ***/
const USER = 0;
const GROUP = 1;
const OTHER = 2;



/**
 * Encode Permission: "rwx" to octal "7"
 * This method expect an array of 3 members i.e. ['rwx', 'rw-', 'r--']
 * will return 764
 * UGO in plain string format is also acceptable
 	@method  toOctal
 	@param {Array|String} UGO
 	@return {Number}
*/
UGOforNODE.prototype.toOctal = function ( UGO ) {
	var toBinaryStageOne, toBinary, digits='';
	
	// if UGO is a String format convert to array
	if ( !Array.isArray( UGO ) ) {
		// slice every 3 characters
		UGO = UGOstringToArray( UGO ); 
	}
	
	// Check if UGO is valid
	if ( !this.isSymbolicValid( UGO ) ) {
		throw new Error( "Permission symbolic representation [U,G,O] is not valid" );
	}
		
		 
	// Convert each UGO section
	UGO.forEach( function ( block ) {
		toBinaryStageOne = block.replace( /[rwx]/ig , 1 );
		toBinary = toBinaryStageOne.replace( /[^1]/g , 0 );
		digits += parseInt( toBinary, 2 )+'';
	});
	return parseInt(digits);
}


/**
 * Encode Octal Permission: "777" to Symbolic notation "rwxrwxrwx"
 * Use octal style without leading "0"
 	@method  toSymbolic
 	@param {Number} octal
 	@param {String} output
 	@return {Array|String}
*/
UGOforNODE.prototype.toSymbolic = function ( octal, output ) {
	var digits, binary, block=[]
		, output = output || 'array';
		
	if ( !this.isOctalValid( octal ) ) {
		throw new Error( "Permission octal representation is not valid" );
	}
		
	
	digits = ( octal ).toString().split('');
	
	digits.forEach( function ( d, index ) {
		var symbole = '';
		binary = (parseInt(d)).toString(2);
		symbole += ( binary >= 100 ) ? 'r' : '-';
		symbole += ( (binary-100) >= 10 ) ? 'w' : '-';
		symbole += ( (binary-110) == 1 ) ? 'x' : '-';
		block[index] = symbole;
	});
	return ( 'string' == output.toLowerCase() ) ? block.join('') : block ;
}



/**
 * Get Permission by Class (User, Group, Other)
	 @method getPermissionByClass
	 @param {Number|String|Array} perm
	 @param {String} user|group|other
	 @return {String}
*/
UGOforNODE.prototype.getPermissionByClass = function ( perm, klass ) {
	var permission = false, result = false;
	
	// is Klass in not defined or not a {String} type
	if ( !klass || !typeof klass == 'string' ) {
		throw new Error( 'Class representation is not valid. Please use: "user", "group", or "other"' );
	}
	
	// Check Octal format
	if ( parseInt( perm ) ) {
		if (!this.isOctalValid(perm))
			throw new Error( "Octal permission representation is not valid" );
		// valid, process
		permission = this.toSymbolic( perm );
	} else {
		// Symbolic format
		if (!this.isSymbolicValid(perm))
			throw new Error( "Octal permission representation is not valid" );
		// Valid, process
		permission = ( Array.isArray( perm ) ) ? perm : UGOstringToArray( perm );
	}
	
	// if Previous steps failed
	if ( !permission )
		throw new Error( "Permission representation is not valid" );
		
	// Select requested class
	switch ( klass.toLowerCase() ) {
		case "user":
			result = permission[USER];
		break;
		case "group":
			result = permission[GROUP];
		break;
		case "other":
			result = permission[OTHER];
		break;
		default:
			throw new Error( 'Class representation is not valid. Please use: "user", "group", or "other"' );
	}
	return result;
}


/**
 * Check Permission validity: "[rwx,rwx,rwx]" to octal "777"
	 @method  isSymbolicValid
	 @param {Array} UGO
	 @return {Boolean}
*/
UGOforNODE.prototype.isSymbolicValid = function ( UGO ) {
	if ( !UGO ) return false;
	if ( !Array.isArray( UGO ) ) UGO = UGOstringToArray(UGO);
	var strJoin = UGO.join('')
		, len = UGO.length
		, strLen = strJoin.length
		, goodChar = !/[^rwx-]/ig.test( strJoin )
	return ( len==3 && strLen==9 && goodChar );
}


/**
 * Check Permission validity: "rwx" to octal "7"
	 @method isOctalValid
	 @param {Number}
	 @return {Boolean}
*/
UGOforNODE.prototype.isOctalValid = function ( octal ) {
	if ( !octal ) return false;
	return !!( parseInt(octal) && octal > 100 && octal < 778 );	
}




/* **************************************************** */
/* 					PRIVATE FUNCTIONS					*/
/* **************************************************** */

/**
 * UGO string to UGO Array
	 @method  UGOstringToArray
	 @param {String} str
	 @return {Boolean}
	 @private
*/
function UGOstringToArray ( str ) {
	if ( str && str.length == 9 ) {
		// slice every 3 characters
		return str.match(/.../g); 
	} else {
		return false;
	}
}